### ISSUE TITLE
Resource design

### ISSUE THREAD

**Issue by rvolosatovs** [wrpc/wrpc#101]
Refs https://github.com/wrpc/wrpc/issues/93 https://github.com/wrpc/wrpc/pull/82

After giving it some thought, trying different things and gathering feedback, I think I finally figured out a reasonable resource design for wRPC.
First, a few key points:
- It is generally expected that wRPC transports are namespaced (and possibly multi-tenant, where relationship of tenants to namespaces is 1:N)
- Resources cannot *escape* the namespace and can only be used within the namespace they originated from.
- Resource handles uniquely identify the origin of the resource and the resource itself in that origin, both of these are application-specific (and may include in itself e.g. the transport namespace)

With this, I think we can untangle the complicated OOP-centric resource design and accept that methods are really just functions, which take borrowed resource handles as first parameters, and they are handled exactly the same as all other functions. (except, perhaps, that for each exported resource, a wRPC peer must eagerly serve `drop` function, which is not explicitly defined in WIT)

In distributed, multi-namespace use cases, the application should be able to derive the namespace that resource originated from using the resource handle. Defining exactly *how* this would work is out-of-scope for wRPC, however wRPC *may* provide utility libraries for building these features e.g. distributed resource tables.

I think that "resources" on the protocol level should just be opaque byte blobs, `list<u8>`, which are entirely application specific

For example, let's consider Wasmtime binary using wRPC for providing host plugins.
Whenever a polyfilled (how about substituted?) resource is received or sent on the wire, it could be encoded as a following WIT record:
```wit
record handle {
   pid: u32,
   id: u32
}
```
- `pid` uniquely identifies the Wasmtime *process*
- `id` uniquely identifies the resource in the WASI context table, i.e. it's just the u32 index, since a single Wasmtime process can only run a single Wasm component

In scenarios where there may be multiple Wasmtime processes simultaneously exporting a particular resource, the pid could be used to figure out which exact Wasmtime process keeps the resource in memory and using some out-of-band, Wasmtime-specific mechanism acquire the resource state.

To enable this, in Rust, `wrpc_transport` crate would define a `Resource` struct roughly as:
```rust
struct Resource<T> { data: Vec<u8>, _ty: PhantomData<T> }
impl<T> AsRef<[u8]> for Resource<T> { .. }
impl<T> From<Vec<u8>> for Resource<T> { .. }
impl<T> From<Resource<T>> for Vec<u8> { .. }
impl<T> Resource<T> { pub fn new(data: Vec<u8>) -> Self { .. } }
```

I'm not sure what's best to do here about the drops, probably for now implementers will just have to manually call `drop` on owned handles, especially since it's potentially a fallible operation. We may want to opt for `Borrow<T>` and `Own<T>` records instead of a singular `Resource<T>`, similar to what's already done for Go https://github.com/wrpc/wrpc/blob/01a536b18c8522469a33cb72a2d5f702585a27ee/go/wrpc.go#L18-L43

For example, for https://github.com/WebAssembly/wasi-keyvalue/ generated Rust`store.open` function signature could look like:
```rust
async fn(identifier: &str) -> Result<wrpc_transport::Own<store::Bucket>, store::Error>
``` 

@jacobvm04 what do you think?

**Comment by rvolosatovs** [comment]
For clarification on the existing design: the resource "roaming" use case was the biggest contributing factor to that path, e.g. the idea of 2 completely different keyvalue store implementations existing in 2 different namespaces and a function being able to operate on both of these simultaneously. A real-world example would be something like:
- Vault keyvalue store in `secrets` namespace
- Redis keyvalue store in `data` namespace
- Some app receiving both a `secrets` namespace bucket and `data` namespace bucket in a single call
 
Omitting the details on how a single component instance was able to acquire handles to two completely different implementations of a keyvalue bucket resource, it should be the platforms'/hosts' responsibility to route the call to the correct namespace - this logic belongs on the application layer and is out-of-scope for wRPC (longer version here https://github.com/wrpc/wrpc/issues/93#issuecomment-2139327912)

**Comment by rvolosatovs** [comment]
> In that case, would the ULID identifying the resources from the current design be moved into the encoding of the resource handle (on an applications-specific basis)? In the "roaming" use case, would the bucket resource handle include an encoding of the origin namespace that the client would decode to decide where to route method calls to?

Yes, indeed, I thing bindgen should not assume any particular format and let the implementers of the interfaces to figure out how to encode and decode resources. For a distributed, multi-namespace, use case, a resource ID could, in fact, be a ULID/UUIDv7 and the "resource table" could be a simple key-value store, e.g. backed by Redis (maybe with some preset TTL for keys), this way every resource could have a globally-unique identifier and potentially even be persistent and resilient to crashes, all while on wRPC/transport level, it's really just a byte blob.

> To put it another way, would the application-specific nature of Resource handles require the clients utilizing them to also use that application specific logic (ex. knowing how to decode a particular resource handle to get the resource's origin namespace)?

I think the clients *may* be aware of the particular resource ID format, but it's probably an anti-pattern. If a client were to create two different resources in two different namespaces, it would need to record the namespace associated with the resource and only call methods within that namespace. It could call a freestanding function, however, in any namespace with these resource handles. In a distributed, multi-namespace, use case the handler of that function (the server) would need to be aware of the resource handle format to extract the origin from it and perhaps call the associated methods in the appropriate namespaces, alternatively, there could be some out-of-band, application-specific logic in place.

**Comment by rvolosatovs** [comment]
I was thinking something like this:

```rust
let client_a = wrpc_transport_http::new_client("example.com")?;
let client_b = wrpc_transport_nats::new_client("myprefix")?;

let bucket_a = get_some_bucket(&client_a, req_context_a, ...).await?; // returns a bucket resource from the "namespace-1" namespace
let bucket_b = get_some_other_bucket(&client_b, req_context_b, ...).await?; // returns a bucket resource from the "namespace-2" namespace
 
bucket_a.set(&client_a, req_context_a, key, value).await?;
bucket_b.set(&client_b, req_context_b, key, value).await?;

bucket_a.set(&client_b, req_context_b, key, value).await?; // not valid in general case, but *may be* depending on the application-specific handling of resources
```

**Comment by jacobvm04** [comment]
Awesome, I see how it works now. Would the next steps then be finishing #94 and then updating #82 to use this approach?

**Comment by rvolosatovs** [comment]
> Awesome, I see how it works now. Would the next steps then be finishing #94 and then updating #82 to use this approach?

I thought we might want to merge https://github.com/wrpc/wrpc/pull/82 first (if only you could rebase on latest `main`, I'll review), then finish up #94 and integrate this design. I think the Wasmtime piece will serve as a reference implementation (currently working on a PR upstream).

I'm planning to get to #94 some time later this week, maybe Wednesday.
btw, if you're interested in working on it, you're more than welcome, and I'm happy to discuss it in more detail! :)

**Comment by jacobvm04** [comment]
That sounds good, I'll work on #82 then! I tried working on #94 a bit last week and some of the encoding/decoding details are a little over my head haha

**Comment by rvolosatovs** [comment]
Giving it a bit more thought, I think what we should just go all-in on resource handle representation be completely application-specific in not define any format for resource handles at all leaving that job for applications.

For example, let's say I wanted to use `wasi:http` with wRPC, but I wanted to avoid the round-trips associated with every method call to e.g. get or set fields on my request/response. As an application developer I could decide to represent my request and response as a "concretely-typed" record, e.g. https://github.com/wrpc/http/blob/5b9a324c74d72b27d8559c54b419a54609d01c68/wit/types.wit#L22-L37, however the byte blob representation would prevent me from nesting async values within the resource representation itself.
To enable such use cases, I think we want to generate a server-side handler trait with associated types for resources (e.g. `Request` and `Response` for HTTP), which would describe how they are to be *decoded* from the byte stream. They would also need to define the paths on which the server may need to subscribe to receive them, similar to what we had with `Subscribe` trait before (https://github.com/wrpc/wrpc/blob/269b164dbcdd70f128da0673dc47260c0fe9f898/crates/transport-legacy/src/lib.rs#L421-L430)

In *outgoing* case it's actually simpler, since `Resource{Own,Borrow}` would simply require the associated type `T` to implement a trait, which describes how it's to be *encoded* (probably using the `T::encoder() -> impl tokio_util::Encode<T>` idea)

**Comment by rvolosatovs** [comment]
So roughly something like:
```rust
trait IncomingResource {
   fn paths() -> &'static [Option<usize>]
   fn read_from(r: impl AsyncRead + Index<[usize]>) -> anyhow::Result<Self>
}

trait OutgoingResource {
   type Encoder: tokio_util::codec::Encoder<Self>
   fn encoder() -> Self::Encoder
}
```

To be fair, these traits are probably generally useful and not specific to resources, so maybe we want to have them implemented for all generated types. I guess that would be the alternative for `Receive` and `Encode` as was discussed in https://github.com/wrpc/wrpc/issues/94#issuecomment-2142914935

**Comment by rvolosatovs** [comment]
Resource handles should really just be small, opaque, byte blobs acting as identifiers.
`wasi:keyvalue/store.bucket` could be represented as a bucket string itself, but for more complex cases, like `wasi:http/types.request`, implementations should define a custom contract for better type safety and composability.
Closing this for now, since I think we now have a reasonable, simple design and an implementation in Rust
